paddle/operators/sequence_concat_op.h

```diff
@@ -0,0 +1,148 @@
+// Case1:
+//  There is one level, the Output LoD will be modified:
+//  LoD(x0) = {{0,2,4}}
+//  LoD(x1) = {{0,1,5}}
+//  LoD(Out) = {{0,3,9}}
+// Case2:
+//  There is two level, and concat level is 1,
+//  the Output LoD will be modified as followed:
+//  LoD(x0) = {{0,2,4}, {0,1,2,3,4}}
+//  LoD(x1) = {{0,3,5}, {0,1,3,4,5}}
+//  LoD(Out) = {{0,5,9}, {0,1,2,4,5,6,7,8,9}}
+template <typename T>
+LoD concatLod(const std::vector<const T*> ins, const size_t axis,
+              const size_t level) {
+  auto out_lod = ins[0]->lod();
+  const size_t n = ins.size();
+  if (axis == 0UL) {
+    if (level == 0) {
+      for (size_t i = 1; i < n; i++) {
+        for (size_t j = 0; j < ins[i]->lod()[0].size(); j++) {
+          out_lod[0][j] += ins[i]->lod()[0][j];
+        }
+      }
+    } else if (level == 1) {
+      for (size_t i = 1; i < n; i++) {
+        PADDLE_ENFORCE_EQ(ins[i]->NumLevels(), 2UL,
+                          "All the LoDTensors of Inputs(X) should "
+                          "have two level.");
+        for (size_t j = 0; j < ins[i]->lod()[0].size(); j++) {
+          out_lod[0].push_back(ins[i]->lod()[0][j]);
+        }
+        for (size_t j = 0; j < ins[i]->lod()[1].size(); j++) {
+          out_lod[1][j] += ins[i]->lod()[1][j];
+        }
+      }
+    }
+  }
+  return out_lod;
+}
+
+template <typename Place, typename T>
+class SequenceConcatOpKernel : public framework::OpKernel {
+ public:
+  void Compute(const framework::ExecutionContext& ctx) const override {
+    auto ins = ctx.MultiInput<LoDTensor>("X");
+    auto* out = ctx.Output<LoDTensor>("Out");
+    const size_t axis = static_cast<size_t>(ctx.Attr<int>("axis"));
+    const size_t level = static_cast<size_t>(ctx.Attr<int>("level"));
+    const size_t n = ins.size();
+    out->mutable_data<T>(ctx.GetPlace());
+    auto out_lod = concatLod<LoDTensor>(ins, axis, level);
+    out->set_lod(out_lod);
+
+    auto out_lod_level = out_lod[level];
+    for (size_t i = 0; i < out_lod_level.size() - 1; i++) {
+      Tensor out_t = out->Slice<T>(static_cast<int>(out_lod_level[i]),
+                                   static_cast<int>(out_lod_level[i + 1]));
+      auto out_stride = framework::stride(out_t.dims());
+      size_t offset = 0;
+
+      for (size_t j = 0; j < n; j++) {
+        auto in_lod_level = ins[j]->lod()[level];
+        auto in_stride = framework::stride(ins[j]->dims());
+        Tensor in_t = ins[j]->Slice<T>(static_cast<int>(in_lod_level[i]),
+                                       static_cast<int>(in_lod_level[i + 1]));
+        size_t axis_dim = in_t.dims()[axis];
+        StridedMemcpy<T>(ctx.device_context(), in_t.data<T>(), in_stride,
+                         in_t.dims(), out_stride, out_t.data<T>() + offset);
+        offset += axis_dim * in_stride[axis];
+      }
+    }
+  }
+};
+
+template <typename Place, typename T>
+class SequenceConcatGradOpKernel : public framework::OpKernel {
```

Comment:
It seems that in the latest dev branch, all the OpKernel should inherit from framework::OpKernel<T> to support different computational accuracies.

Reply:
Done.